subprojects/core-api/src/main/java/org/gradle/api/artifacts/ComponentMetadataDetails.java

```diff
@@ -63,4 +64,13 @@
      * @since 4.4
      */
     void withVariant(String name, Action<VariantMetadata> action);
+
+    /**
+     * Adjust the component-level attributes.
+     *
+     * @param action the adjustment action to be executed on attributes
+     *
+     * @since 4.5
+     */
+    void withAttributes(Action<? super AttributeContainer> action);
```

Comment:
:x: Should also be able to read attributes via `ComponentMetadata` supertype. Then the attributes can be consumed in `ComponentSelectionRules`

Reply:
Right, I wasn't sure if we wanted to expose such an API there. I will add it.